core/base/abstract_factory.hpp

```diff
@@ -0,0 +1,246 @@
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_BASE_ABSTRACT_FACTORY_HPP_
+#define GKO_CORE_BASE_ABSTRACT_FACTORY_HPP_
+
+
+#include "core/base/polymorphic_object.hpp"
+
+
+namespace gko {
+
+
+/**
+ * The AbstractFactory is a generic interface template that enables easy
+ * implementation of the abstract factory design pattern.
+ *
+ * The interface provides the AbstractFactory::generate() method that can
+ * produce products of type `AbstractProductType` using an object of
+ * `ComponentsType` (which can be constructed on the fly from parameters to its
+ * constructors).
+ * The generate() method is not declared as virtual, as this allows subclasses
+ * to hide the method with a variant that preserves the compile-time type of the
+ * objects. Instead, implementers should override the generate_impl() method,
+ * which is declared virtual.
+ *
+ * Implementers of concrete factories should consider using the
+ * EnableDefaultFactory mixin to obtain default implementations of utility
+ * methods of PolymorphicObject and AbstractFactory.
+ *
+ * @tparam AbstractProductType  the type of products the factory produces
+ * @tparam ComponentsType  the type of components the factory needs to produce
+ *                         the product
+ */
+template <typename AbstractProductType, typename ComponentsType>
+class AbstractFactory
+    : public EnableAbstractPolymorphicObject<
+          AbstractFactory<AbstractProductType, ComponentsType>> {
+public:
+    using abstract_product_type = AbstractProductType;
+    using components_type = ComponentsType;
+
+    /**
+     * Creates a new product from the given components.
+     *
+     * The method will create an ComponentsType object from the arguments of
+     * this method, and pass it to the generate_impl() function which will
+     * create a new AbstractProductType.
+     *
+     * @tparam Args  types of arguments passed to the constructor of
+     *               ComponentsType
+     *
+     * @param args  arguments passed to the constructor of ComponentsType
+     *
+     * @return an instance of AbstractProductType
+     */
+    template <typename... Args>
+    std::unique_ptr<AbstractProductType> generate(Args &&... args) const
```

Comment:
It is (almost) always a better idea to return a `unique_ptr` than a `shared_ptr`. A (rvalue) `unique_ptr` can always be implicitly converted into a `shared_ptr`, and the opposite is not true.
Thus, the code `gko::shared(factory->generate(args...))` will always get you a `shared_ptr`.

The only reason why a factory would return a shared pointer is if one of the implementations of `AbstractProductType` uses the `std::enable_shared_from_this` mixin, which is the case for  `CpuExecutor`. In that case the object has to be constructed and assigned to a `shared_ptr` in order for `shared_from_this()` method to work correctly. 

I don't think this will be a common use-case, so I would not overcomplicate the design now, and add this possibility later if the need arises.